cpp/src/fil/fil.cu

```diff
@@ -325,11 +320,33 @@ struct forest {
   float global_bias_ = 0;
   shmem_size_params class_ssp_, proba_ssp_;
   int fixed_block_count_ = 0;
+  int max_shm_           = 0;
   // Optionally used
   rmm::device_uvector<float> vector_leaf_;
   cat_sets_device_owner cat_sets_;
 };
 
+template <typename storage_type>
+struct enable_smem_carveout : dispatch_functor<void> {
+  const int MAX_SHM_STD = 48 * 1024;  // maximum architecture-independent size
+  const int max_shm;
+  enable_smem_carveout(int max_shm_) : max_shm(max_shm_) {}
+
+  template <class KernelParams = KernelTemplateParams<>>
+  void run(predict_params p)
+  {
+    void (*kernel)(storage_type, predict_params) = infer_k<KernelParams::N_ITEMS,
```

Comment:
I wonder if it would be better if `infer_k` just accepts `KernelParams` as a template parameter. This change would be for a separate pull request, of course.